lib/init-7.scm

```diff
@@ -820,13 +820,10 @@
             ((null? p) (list _and (list _null? v) (k vars)))
             (else (list _and (list _equal? v p) (k vars))))))))
     (define ellipsis-mark?
-      (if (if ellipsis-specified?
-              (memq ellipsis lits)
-              (any (lambda (x) (compare ellipsis x)) lits))
-          (lambda (x) #f)
-          (if ellipsis-specified?
-              (lambda (x) (eq? ellipsis x))
-              (lambda (x) (compare ellipsis x)))))
+      (let ((cmp (if ellipsis-specified? eq? compare)))
+        (if (any (lambda (x) (cmp ellipsis x)) lits)
+            (lambda (x) #f)
+            (lambda (x) (cmp ellipsis x)))))
```

Comment:
This is slower in the common case, I wouldn't bother with the refactoring you did here.